In [4]:
import random

In [5]:
def average_py(n):
    s = 0
    for i in range(n):
        s += random.random()
    return s / n 

In [6]:
n = 10000000

In [7]:
%time average_py(n)

CPU times: total: 1.77 s
Wall time: 1.84 s


0.5001361958846964

In [8]:
%timeit average_py(n)

2.4 s ± 31.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [9]:
%time sum([random.random() for _ in range(n)]) / n 

CPU times: total: 3.59 s
Wall time: 3.71 s


0.49997722083384144

<span style='font-size:17px'><b>Basic Python Benchmarking</b>

* <b>Function Definition:</b> Creates a pure Python function 'average_py' that uses a simple 'for' loop to calculate the mean of $n$ random numbers.
* <b>Basline Performance:</b> Benchmarks the function with 10 million iterations, recording a wall time of 1.31 seconds for a single execution.
* <b>Statistical Timing:</b> Uses '%timeit' to provide a more robust average, showing approximately 1.36 seconds per loop across 7 different runs.
* <b>List Comprehension Comparison:</b> Tests a one-liner approach using 'sum()' on a list comprehension, which performs slower at 2.17 seconds due to memory overhead.

In [10]:
import numpy as np

In [12]:
def average_np(n):
    s = np.random.random(n)
    return s.mean()


In [13]:
%time average_np(n)

CPU times: total: 281 ms
Wall time: 287 ms


np.float64(0.5001041588030305)

In [14]:
%timeit average_np(n)

307 ms ± 21.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [15]:
s = np.random.random(n)
s.nbytes

80000000

<span style='font-size:17px'><b>NumPy Vectorization</b>

* <b>Function Optimization:</b> Implements 'average_np' using NumPy's 'random' and 'mean' methods to replace the manual Python loop.
* <b>Syntax Debugging:</b> Identifies that 's.mean' returns the function object whereas 's.mean()' is required to execute the calculation.
* <b>Performance Leap:</b> Acheives a wall time of 133 ms for 10 million numbers, making it roughly 10x faster than the pure Python loop.
* <b>Vectorized Execution:</b> Demonstrates how NumPy processes the entire arra at once in C, drastically reducing the overhead compared to Python iteration.

In [16]:
import numba

In [17]:
average_nb = numba.jit(average_py)

In [18]:
%time average_nb(n)

CPU times: total: 2.33 s
Wall time: 3.88 s


0.5000224719455327

In [19]:
%time average_nb(n)

CPU times: total: 188 ms
Wall time: 207 ms


0.5000588154115762

In [20]:
%timeit average_nb(n)

205 ms ± 9.04 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


<span style='font-size:17px'><b>Numba Just-In-Time (JIT) Compilation</b>

* <b>Compilation Overhead:</b> Incurs a one-time 'warm-up' cost during the first execution as Numba translates Python bytecode into optimized machine code.
* <b>Execution Efficiency:</b> Reduces execution time from ~1.3 seconds to 119 milliseconds once the function is cached in its compiled state.
* <b>Seamless Integration:</b> Demonstrates the 'numba.jit()' wrapper's ability to take a standard Python function and optimize it without manual code rewrites.
* <b>Performance Parity:</b> Acheives results comparable to NumPy's vectorized methods by running calculations directly at the CPU instruction level.

!pip install Cython

In [21]:
%load_ext Cython

In [22]:
%%cython -a 
import random
def average_cy1(int n):
    cdef int i
    cdef float s = 0 
    for i in range(n):
        s += random.random()
    return s/n

In [23]:
%time average_cy1(n)

CPU times: total: 3.05 s
Wall time: 3.12 s


0.5000942945480347

In [24]:
%timeit average_cy1(n)

3.13 s ± 50.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [25]:
%%cython 
from libc.stdlib cimport rand 
cdef extern from 'limits.h':
    int INT_MAX
cdef int i
cdef float rn
for i in range(5):
    rn = rand() / INT_MAX
    print(rn)

1.2555159628391266e-05
1.1391937732696533e-05
2.6565976440906525e-06
1.3106036931276321e-05
1.0841060429811478e-05


In [26]:
%%cython -a 
from libc.stdlib cimport rand
cdef extern from 'limits.h':
    int INT_MAX
def average_cy2(int n):
    cdef int i 
    cdef float s = 0
    for i in range(n):
        s += rand() / INT_MAX
    return s / n

In [27]:
%time average_cy2(n)

CPU times: total: 328 ms
Wall time: 326 ms


7.628993898833869e-06

In [28]:
%timeit average_cy2(n)

297 ms ± 39.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


<span style='font-size:17px'><b>Cython and C-Library Integration</b>

* <b>Library Transition:</b> Shifts from the high-level Python 'random' module to the low-level 'libc.stdlib' C library to eliminate interpreter overhead.
* <b>Static Declaration:</b> Uses 'cdef' to explicitly define variable types and external C constants like 'INT_MAX', allowing for direct machine-code mapping.
* <b>Optimization Analysis:</b> Employs the '-a' annotation flag to identify and remove 'yellow' Python-interaction lines, achieving a 'white' pure-C execution loop.
* <b>Efficiency Breakthrough:</b> Reduces execution time from ~2.0 seconds to 153 milliseconds by bypassing the Python virtual machine entirely during the calculation.

In [29]:
def is_prime(I):
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True


In [30]:
n = int(1e8 + 3)
n

100000003

In [31]:
%time is_prime(n)

CPU times: total: 15.6 ms
Wall time: 526 μs


False

In [32]:
p1 = int(1e8 + 7)
p1

100000007

In [33]:
%time is_prime(p1)

CPU times: total: 0 ns
Wall time: 1.19 ms


True

In [34]:
p2 = 100109100129162907

In [35]:
p2.bit_length()

57

In [36]:
%time is_prime(p2)

CPU times: total: 37.6 s
Wall time: 39.4 s


True

<span style='font-size:17px'><b>Computational Primality Testing & Efficiency</b>

* <b>Optimized Trial Division:</b> The 'is_prime' function implements a loop that only checks odd numbers up to the square root of the target ($\sqrt{1}$), which is a significantly more efficient approach than checking every number up to $I$.
* <b>Time Complexity Practice:</b> The results show a massive jump in 'Wall Time' as the numbers grow; while $10^8 + 3$ is processed in **$68.4 \ \ mu s$**, the 57-bit prime $p2$ requires **$15$ seconds** to verify.
* <b>Bit-Length Impact:</b> The use of 'p2.bit_length()' highlights that asthe binary size of a number increases (to **57 bits** in this case), the number of potential divisors grows, causing a linear algorithm to slow down noticeably.
* <b>Performance Monitoring:</b> By using the '%time' magic command, the output distinguishes **CPU time** (active processing) and **Wall time** (total time elapsed), which is identifying bottlenecks in cryptographic or mathematical scripts.

In [37]:
is_prime_nb = numba.jit(is_prime)

In [38]:
%time is_prime_nb(n)

CPU times: total: 328 ms
Wall time: 369 ms


False

In [39]:
%time is_prime_nb(n)

CPU times: total: 0 ns
Wall time: 24.1 μs


False

In [40]:
%time is_prime_nb(p1)

CPU times: total: 0 ns
Wall time: 65.1 μs


True

In [41]:
%time is_prime_nb(p2)

CPU times: total: 2.02 s
Wall time: 2.05 s


True

In [42]:
%%cython
def is_prime_cy1(I):
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True

In [43]:
%timeit is_prime(p1)

1.14 ms ± 24.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [44]:
%timeit is_prime_cy1(p1)

748 μs ± 32 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [45]:
%%cython
def is_prime_cy2(long I):
    cdef long i
    if I % 2 == 0: return False
    for i in range(3, int(I ** 0.5) + 1, 2):
        if I % i == 0: return False
    return True

In [46]:
%timeit is_prime_cy2(p1)

47.7 μs ± 907 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [47]:
%timeit is_prime_nb(p1)

50.1 μs ± 1.59 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


<span style='font-size:17px'><b>Performance Scaling and Limitations in Cython and Numba</b>

* <b>Static Typing Efficiency:</b> The transition from 'is_prime_cy1' to 'is_prime_cy2' demonstrates that explicitly declaring types (e.g., 'long I' and 'cdef long i') allows Cython to bypass Python's overhead, reducing execution time for $p1$ from **$334 \ \mu s$** results likely utilizes a 64-bit environment.
* <b>The 32-bit vs. 64-bit Bottleneck:</b> The 'OverflowError' encounterd with $p2$ occurs because the code attempts to fit a **57-bit** integer into a C 'long'; on Windows systems, this type is typically limited to 32 bits, whereas the book's successful **1.68s** result likely utilizes a 64-bit environment.
* <b>Numba's JIT advantage:</b> Numba provides a high-performance alternative to Cython with less manual configuration, successfully processing the large prime $p2$ in **$657 \ ms$** compared to the original Python time of **$15 \ seconds$**.
* <b>Compilation Latency:</b> Initial calls to Numba-decorated functions (like the first run of 'is_prime_nb') includes a one-time 'warm-up' cost for machine code compilation (**$192 \ ms$**), while subsequent calls run at near-native speeds (**$11.2 \ \mu s$**).

In [48]:
def fib_rec_py1(n):
    if n < 2: 
        return n
    else:
        return fib_rec_py1(n - 1) + fib_rec_py1(n - 2)

In [49]:
%time fib_rec_py1(35)

CPU times: total: 6.33 s
Wall time: 7.38 s


9227465

In [50]:
import numba

@numba.jit(nopython=True)
def fib_rec_nb(n):
    if n < 2:
        return n
    else:
        return fib_rec_nb(n - 1) + fib_rec_nb(n - 2)

%time fib_rec_nb(35)

CPU times: total: 500 ms
Wall time: 518 ms


9227465

In [51]:
%%cython
def fib_rec_cy(int n):
    if n < 2:
        return n
    else:
        return fib_rec_cy (n - 1) + fib_rec_cy(n -2)

In [52]:
%time fib_rec_cy(35)

CPU times: total: 3.39 s
Wall time: 3.49 s


9227465

In [53]:
from functools import lru_cache as cache

In [54]:
@cache(maxsize=None)
def fib_rec_py2(n):
    if n < 2:
        return n 
    else:
        return fib_rec_py2(n -1) + fib_rec_py2(n -2)

In [55]:
%time fib_rec_py2(35)

CPU times: total: 0 ns
Wall time: 58.7 μs


9227465

In [56]:
%time fib_rec_py2(80)

CPU times: total: 0 ns
Wall time: 70.1 μs


23416728348467685

<span style='font-size:17px'><b>The Performance Toolkit: Libraries & Logic</b>

* <b>Python (The Foundation):</b> Standard Python is flexible but slow because it handles one task at a time and spends a lot of 'brainpower' managing data types.
* <b>Cython & Numba (The Speed Boosters):</b> These act like translators that turn slow Python instructions into high-speed machine code, making itself run up to 100x faster.
* <b>Multiprocessing (The Teamwork):</b> This doesn't make the code faster; it just hires 'more workers' (CPU cores)so you can finish a long list of tasks simulateously.
* <b>Functools/Caching (The Memory):</b> While the others focus on raw speed, caching fixes the logic by letting the computer remember its answer so it never repeats the same work twice.

In [57]:
def fib_it_py(n):
    x, y = 0, 1
    for i in range(1, n +1):
        x, y = y, x + y
    return x 

In [58]:
%time fib_it_py(80)

CPU times: total: 0 ns
Wall time: 36 μs


23416728348467685

In [59]:
fib_it_nb = numba.jit(fib_it_py)

In [60]:
%time fib_it_nb(80)

CPU times: total: 234 ms
Wall time: 303 ms


23416728348467685

In [61]:
%time fib_it_nb(80)

CPU times: total: 0 ns
Wall time: 29.3 μs


23416728348467685

In [62]:
%%cython
def fib_it_cy1(int n):
    cdef long i
    cdef long x = 0, y = 1
    for i in range(1, n + 1):
        x, y = y, x + y 
    return x   

In [63]:
%time fib_it_cy1(80)

CPU times: total: 0 ns
Wall time: 37.4 μs


-285007387

In [64]:
%%time
fn = fib_rec_py2(150)
print(fn)

9969216677189303386214405760200
CPU times: total: 0 ns
Wall time: 385 μs


In [65]:
fn.bit_length()

103

In [66]:
%%time 
fn = fib_it_nb(150)
print(fn)

6792540214324356296
CPU times: total: 0 ns
Wall time: 281 μs


In [67]:
fn.bit_length()

63

In [68]:
%%cython
def fib_it_cy_final(int n):
    cdef int i
    cdef object x = 0 
    cdef object y = 1
    
    for i in range(1, n + 1):
        x, y = y, x + y
        
    return x

Content of stdout:
_cython_magic_a37517bacbd84a36f704482713e4188dbf806b2ae9aeaeaf95bb1ac4990cb6b5.c
   Creating library C:\Users\wyatt\.ipython\cython\_cython_magic_a37517bacbd84a36f704482713e4188dbf806b2ae9aeaeaf95bb1ac4990cb6b5.cp313-win_amd64.lib and object C:\Users\wyatt\.ipython\cython\_cython_magic_a37517bacbd84a36f704482713e4188dbf806b2ae9aeaeaf95bb1ac4990cb6b5.cp313-win_amd64.exp
Generating code
Finished generating code

In [69]:
%%time
fn = fib_it_cy2(150)
print(fn)

NameError: name 'fib_it_cy2' is not defined

In [70]:
fn.bit_length()

63

<span style='font-size:17px'><b>The Iterative Loop: Speeding into a Wall</b>

* <b>Logic Efficiency:</b> By switching to a Iterative Algorithm ('for i in range'), the 'forgetfulness' of the recursion is fixed. The computer nows only does $N$ steps of math instead of millions, which is why the time dropped to microseconds.
* <b>The 'Box' Problem:</b> While the loop is fast, the variables ($x$ and $y$) are like container. In Cython and Numba, these containers are high-speed '64-bit boxes' that can only hold a certain amount of data.
* <b>The 63-Bit Overflow:</b> Once your iterative loop reaches $Fib(93)$, the number bceomes too large for a 64-bit box. The extra data 'spills over,' causing the bits to flip and creating the negative numbers or 63-bit cutoff seen in the results.
* <b>Why Python Wins:</b> Standard Python's iterative loop is slightly slower because it uses 'Expanding Tanks' instead of 'Fixed Boxes.' It waits for the number to grow and automatically gives it more memory, which is how it reaches 103 bits for $Fib(150)$.

import random 
import numpy as np
from pylab import mpl, plt
plt.style.use('seaborn-v0_8')
mpl.rcParams['font.family'] = 'serif'
%matplotlib inline

In [71]:
rn = [(random.random() * 2 - 1, random.random() * 2 - 1)
      for i in range(500)]

In [72]:
rn = np.array(rn)
rn[:5]

array([[-0.37844156, -0.82368639],
       [-0.08600534, -0.41261614],
       [ 0.33801654, -0.09895997],
       [-0.99246109,  0.20434203],
       [-0.9071525 , -0.43151697]])

In [76]:
fig = plt.figure(figsize=(7, 7))
ax = fig.add_subplot(1, 1, 1)
circ = plt.Circle((0,0), radius=1, edgecolor = 'g', lw=2.0,
                  facecolor='None')
box = plt.Rectangle((-1, 1), 2, 2, edgecolor='b', alpha=0.3)
ax.add_patch(box)
ax.add_patch(circ)
plt.plot(rn[:, 0], rn [:, 1], 'r.')
plt.ylim(-1.1, 1.1)
plt.xlim(-1.1, 1.1)
plt.show()

NameError: name 'plt' is not defined

In [77]:
n = int(1e7)

In [78]:
%time rn = np.random.random((n, 2)) * 2 -1

CPU times: total: 547 ms
Wall time: 548 ms


In [79]:
rn.nbytes

160000000

In [80]:
%time distance = np.sqrt((rn ** 2).sum(axis=1))
distance[:8].round(3)

CPU times: total: 562 ms
Wall time: 553 ms


array([0.745, 1.375, 0.783, 0.75 , 0.675, 0.906, 0.881, 0.799])

In [81]:
%time frac = (distance <= 1.0).sum() / len(distance)

CPU times: total: 31.2 ms
Wall time: 31.6 ms


In [82]:
pi_mcs = frac * 4
pi_mcs

np.float64(3.1414552)

In [83]:
def mcs_pi_py(n):
    circle = 0
    for _ in range(n):
        x, y = random.random(), random.random()
        if (x ** 2 + y ** 2) ** 0.5 <= 1:
            circle += 1
    return (4 * circle) / n 

In [84]:
%time mcs_pi_py(n)

CPU times: total: 11.9 s
Wall time: 12.2 s


3.14125

In [85]:
mcs_pi_nb = numba.jit(mcs_pi_py)

In [86]:
%time mcs_pi_nb(n)

CPU times: total: 1.06 s
Wall time: 1.11 s


3.1418976

In [87]:
%time mcs_pi_nb(n)

CPU times: total: 406 ms
Wall time: 395 ms


3.1421252

In [88]:
%%cython
from libc.math cimport sqrt  
import random

def mcs_pi_cy1(int n):
    cdef int i, circle = 0
    cdef double x, y
    for i in range(n):
        x = random.random()
        y = random.random()
        if sqrt(x*x + y*y) <= 1.0:
            circle += 1
    return (4.0 * circle) / n

Content of stdout:
_cython_magic_2bce9bfc47c829b845dc8ea84a05e2e1be82405fef541fca18b5f0a908705ea0.c
   Creating library C:\Users\wyatt\.ipython\cython\_cython_magic_2bce9bfc47c829b845dc8ea84a05e2e1be82405fef541fca18b5f0a908705ea0.cp313-win_amd64.lib and object C:\Users\wyatt\.ipython\cython\_cython_magic_2bce9bfc47c829b845dc8ea84a05e2e1be82405fef541fca18b5f0a908705ea0.cp313-win_amd64.exp
Generating code
Finished generating code

In [89]:
%time mcs_pi_cy1(n)

CPU times: total: 4.62 s
Wall time: 5.05 s


3.1406216

In [90]:
%%cython -a
from libc.stdlib cimport rand, RAND_MAX
from libc.math cimport sqrt

def mcs_pi_cy2(int n):
    cdef int i, circle = 0
    cdef double x, y
    for i in range(n):
        # The <double> here is the magic fix
        x = rand() / <double>RAND_MAX
        y = rand() / <double>RAND_MAX
        if sqrt(x*x + y*y) <= 1.0:
            circle += 1
    return (4.0 * circle) / n

In [91]:
%time mcs_pi_cy2(n)

CPU times: total: 531 ms
Wall time: 537 ms


3.1411636

<span style='font-size:17px'><b>The Monte Carlo Pi Simulation Breakdown</b>

* <b>The Geometric Core:</b> The algorithm estimates the $\pi$ by randomly 'throwing darts' at a unit square (area of 1) and checking how many land inside a quarter-circle arc (are of $\pi/4$). By calculating the ratio of 'hits' in the circle to the total number of darts thrown and multiplying by 4, we arrive at an estimate of $\pi$.
* <b>The Python Performance Gap:</b> A standard Python loop ('mcs_pi_py') is slow because the interpreter must manage high-level objects for every single iteration, which can lead to execution times of several seconds for 10 million loops. Using NumPy or Numba speeds this up significantly by performing operations in bulk or compiling them into machine code, dropping the runtime to under 600 milliseconds.
* <b>The Cython 'C-speed' Secret:</b> In your Cython version ('mcs_pi_cy2'), the code is transformed into pure C, which eliminates the heavy Python overhead. By using 'cimport' to access the C library's 'rand()' and 'sqrt()' functions and explicitly declaring variables as 'int' or 'double', the loop runs at hardware-level speeds (roughly 359 milliseconds).


In [92]:
import math

In [93]:
S0 = 36.
T = 1.0
r = 0.06
sigma = 0.2

In [94]:
def simulate_tree(M):
    dt = T/M
    u = math.exp(sigma * math.sqrt((dt)))
    d = 1 / u
    S = np.zeros((M +1, M + 1))
    S[0,0] = S0
    z = 1
    for t in range(1, M + 1):
        for i in range(z):
            S[i, t] = S[i, t-1] * u
            S[i+1, t] = S[i, t-1] * d 
        z += 1 
    return S

In [95]:
np.set_printoptions(formatter={'float':
                               lambda x: '%6.2f' % x}) 

In [96]:
simulate_tree(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]])

In [97]:
%time simulate_tree(500)

CPU times: total: 234 ms
Wall time: 238 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [  0.00,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [  0.00,   0.00,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]])

In [98]:
M = 4

In [99]:
up = np.arange(M + 1)
up = np.resize(up, (M + 1, M + 1))
up

array([[0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4],
       [0, 1, 2, 3, 4]])

In [100]:
down = up.T * 2 
down

array([[0, 0, 0, 0, 0],
       [2, 2, 2, 2, 2],
       [4, 4, 4, 4, 4],
       [6, 6, 6, 6, 6],
       [8, 8, 8, 8, 8]])

In [101]:
up - down

array([[ 0,  1,  2,  3,  4],
       [-2, -1,  0,  1,  2],
       [-4, -3, -2, -1,  0],
       [-6, -5, -4, -3, -2],
       [-8, -7, -6, -5, -4]])

In [102]:
dt = T / M 

In [103]:
S0 * np.exp(sigma * math.sqrt(dt) * (up - down))

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [ 29.47,  32.57,  36.00,  39.79,  43.97],
       [ 24.13,  26.67,  29.47,  32.57,  36.00],
       [ 19.76,  21.84,  24.13,  26.67,  29.47],
       [ 16.18,  17.88,  19.76,  21.84,  24.13]])

In [104]:
def simulate_tree_np(M):
    dt = T / M
    up = np.arange(M + 1)
    up = np.resize(up, (M + 1, M + 1))
    down = up.transpose() * 2
    S = S0 * np.exp(sigma * math.sqrt(dt) * (up-down))
    return S

In [105]:
simulate_tree_np(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [ 29.47,  32.57,  36.00,  39.79,  43.97],
       [ 24.13,  26.67,  29.47,  32.57,  36.00],
       [ 19.76,  21.84,  24.13,  26.67,  29.47],
       [ 16.18,  17.88,  19.76,  21.84,  24.13]])

In [106]:
%time simulate_tree_np(500)

CPU times: total: 15.6 ms
Wall time: 17.6 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [ 35.36,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [ 34.73,  35.05,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.41,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.40,   0.41,   0.41]])

In [107]:
simulate_tree_nb = numba.jit(simulate_tree)

In [108]:
simulate_tree_nb(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]])

In [109]:
%time simulate_tree_nb(500)

CPU times: total: 0 ns
Wall time: 1.89 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [  0.00,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [  0.00,   0.00,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]])

In [110]:
%timeit simulate_tree_nb(500)

2.6 ms ± 242 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [111]:
%%cython -a
import numpy as np
cimport numpy as np
from libc.math cimport exp, sqrt

def simulate_tree_cy(int M):
    cdef double S0 = 36.0
    cdef double T = 1.0
    cdef double sigma = 0.2
    
    cdef int z, t, i
    cdef double dt, u, d
    cdef double[:, :] S = np.zeros((M + 1, M + 1), dtype=np.float64)

    dt = T / M
    u = exp(sigma * sqrt(dt))
    d = 1.0 / u
    S[0, 0] = S0
    z = 1
    
    for t in range(1, M + 1):
        for i in range(z):
            S[i, t] = S[i, t-1] * u
            S[i+1, t] = S[i, t-1] * d
        z += 1
        
    return np.array(S)

Content of stdout:
_cython_magic_a24e1f3e4251966382b4b9c0e5545c9b2e6f0d738d78c44d154086525c8374b0.c
C:\Users\wyatt\anaconda3\Lib\site-packages\numpy\_core\include\numpy\npy_1_7_deprecated_api.h(14) : Warning Msg: Using deprecated NumPy API, disable it with #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
   Creating library C:\Users\wyatt\.ipython\cython\_cython_magic_a24e1f3e4251966382b4b9c0e5545c9b2e6f0d738d78c44d154086525c8374b0.cp313-win_amd64.lib and object C:\Users\wyatt\.ipython\cython\_cython_magic_a24e1f3e4251966382b4b9c0e5545c9b2e6f0d738d78c44d154086525c8374b0.cp313-win_amd64.exp
Generating code
Finished generating code

In [112]:
simulate_tree_cy(4)

array([[ 36.00,  39.79,  43.97,  48.59,  53.71],
       [  0.00,  32.57,  36.00,  39.79,  43.97],
       [  0.00,   0.00,  29.47,  32.57,  36.00],
       [  0.00,   0.00,   0.00,  26.67,  29.47],
       [  0.00,   0.00,   0.00,   0.00,  24.13]])

In [113]:
%time simulate_tree_cy(500)

CPU times: total: 0 ns
Wall time: 5.1 ms


array([[ 36.00,  36.32,  36.65, ..., 3095.69, 3123.50, 3151.57],
       [  0.00,  35.68,  36.00, ..., 3040.81, 3068.13, 3095.69],
       [  0.00,   0.00,  35.36, ..., 2986.89, 3013.73, 3040.81],
       ...,
       [  0.00,   0.00,   0.00, ...,   0.42,   0.42,   0.43],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.41,   0.42],
       [  0.00,   0.00,   0.00, ...,   0.00,   0.00,   0.41]])

In [114]:
%timeit S = simulate_tree_cy(500)

6.06 ms ± 485 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


<span style='font-size:17px'><b>The High-Performance Binomial Tree Breakdown</b>

* <b>The Vectorization Shift:</b> While the original 'simulate_tree' used nested loops, the NumPy version ('simulate_tree_np') uses broadcasting and matrix math to calculate every possible price at once. By replacing the loops with 'up - down' index logic, the execution for 500 steps dropped from <b>167 ms</b> to a staggering <b>11.5ms</b>.
* <b>The Numba 'Magic':</b> By simply wrapping the original function in the '@numba.jit' decorator, the computer was able to compile the Python loops into machine code. This provides a massive speed boost (reducing wall time to under 1 ms) without having to rewrite a single line of the original logic.
* <b>Cython's Static Typing:</b> In the 'simulate_tree_cy' version, we explicitly declared variables like 'cdef double S0' and 'cdef int z, t, i'. This tells the computer exactly how much memory to set aside upfront, allowing it to bypass the slow 'type-checking' that makes standard Python sluggish.
* <b>C-Standard Math Integration:</b> By importing 'exp' and 'sqrt' directly from 'libc.math', the simulation avoids calling high-level Python math objects. This ensures that the most computationally expensive operations are performed using optimized C libraries, resulting in a line 4.61 ms runtime.
* <b>Efficient Recombining Structure:</b> The algorithm utilizes a 'recombining' tree where an upward move followed by a downward move returns to the same point. This prevents the number from exploding exponentially, ensuring the model remains computationally scalable even as the number of time steps increases.


In [115]:
M = 100
I = 50000

In [116]:
def mcs_simulation_py(p):
    M, I = p
    dt = T / M
    S = np.zeros((M + 1, I))
    S[0] = S0
    rn = np.random.standard_normal(S.shape)
    for t in range(1, M + 1):
        for i in range(I):
            S[t, i] = S[t-1, i]* math.exp((r - sigma ** 2/ 2) * dt+
                                          sigma * math.sqrt(dt) * rn[t, i])
    return S

In [117]:
%time S = mcs_simulation_py((M, I))

CPU times: total: 13.3 s
Wall time: 14.8 s


In [118]:
S[-1].mean()

np.float64(38.233067700034006)

In [119]:
S0 * math.exp(r * T)

38.22611567563295

In [120]:
K = 40

In [121]:
C0 = math.exp(-r * T) * np.maximum(K -S[-1], 0).mean()

In [122]:
C0  #

np.float64(3.8281976108262956)

<span style='font-size:17px'><b>The Euler Scheme: Discretization & Implementation</b>

* <b>The Difference Equation:</b> The code implements the Euler Scheme to approximate the Continuous-Time BSM equation by breaking it into discrete time steps ($\Delta t$). This allows the computer to iteratively calculate the next stock price by adding a fixed drift and a random stochastic shock to the current price.
* <b>Log-Normal Distribution Correction:</b> By using the exponential form $e^{(r - 0.5 \ sigma^2)}$,the simulation ensures that the stock price remains positive and follows a log-normal distribution. The $-0.5\ sigma^2$ term is a critical adjustment by Ito's Lemma to ensure the expected value of the simulation matches the theoretical risk-neutral growth.
* <b>Payoff Valuation:</b> The final calculation of $C0$ (3.85) demonstrates the power of the Monte Carlo method under the Euler Scheme. By taking the average of all terminal payoffs and discounting them back to the present at the risk-free rate, the model provides an accurate numerical estimate of the option's fair market value.

In [123]:
mcs_simulation_nb = numba.jit(mcs_simulation_py)

In [124]:
%time S = mcs_simulation_nb((M, I))

CPU times: total: 2.45 s
Wall time: 2.56 s


In [125]:
%time S = mcs_simulation_nb((M,I))

CPU times: total: 500 ms
Wall time: 617 ms


In [126]:
S[-1].mean()

np.float64(38.25462731583732)

In [127]:
C0 = math.exp(-r * T) * np.maximum(K - S[-1], 0).mean()

In [128]:
C0

np.float64(3.8335634468696917)

In [129]:
%timeit S = mcs_simulation_nb((M, I))

451 ms ± 8.23 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [137]:
%%cython
import numpy as np
cimport numpy as np
cimport cython
from libc.math cimport exp, sqrt
cdef float S0 = 36.
cdef float T = 1.0
cdef float r = 0.06
cdef float sigma = 0.2
@cython.wraparound(False)
def mcs_simulation_cy(p):
    cdef int M,I
    M, I = p 
    cdef float dt = T / M
    cdef double [:, :] S = np.zeros((M + 1, I))
    cdef double [:, :] rn = np.random.standard_normal((M + 1, I))
    cdef int t, i
    S[0]
    for t in range(1, M + 1):
        for i in range(I):
            S[t, i] = S[t-1, i] * exp((r - sigma ** 2/2) * dt +
                                      sigma * sqrt(dt) * rn[t, i])
    return np.array(S)         

Content of stdout:
_cython_magic_67f323e7d72d79f80d7e63fb8e6903c729ec5be2bbd40c2cdd838ace4523083e.c
C:\Users\wyatt\anaconda3\Lib\site-packages\numpy\_core\include\numpy\npy_1_7_deprecated_api.h(14) : Warning Msg: Using deprecated NumPy API, disable it with #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
C:\Users\wyatt\.ipython\cython\_cython_magic_67f323e7d72d79f80d7e63fb8e6903c729ec5be2bbd40c2cdd838ace4523083e.c(21264): warning C4305: '=': truncation from 'double' to 'float'
C:\Users\wyatt\.ipython\cython\_cython_magic_67f323e7d72d79f80d7e63fb8e6903c729ec5be2bbd40c2cdd838ace4523083e.c(21273): warning C4305: '=': truncation from 'double' to 'float'
   Creating library C:\Users\wyatt\.ipython\cython\_cython_magic_67f323e7d72d79f80d7e63fb8e6903c729ec5be2bbd40c2cdd838ace4523083e.cp313-win_amd64.lib and object C:\Users\wyatt\.ipython\cython\_cython_magic_67f323e7d72d79f80d7e63fb8e6903c729ec5be2bbd40c2cdd838ace4523083e.cp313-win_amd64.exp
Generating code
Finished generating code

In [138]:
%time S = mcs_simulation_cy((M, I))

CPU times: total: 953 ms
Wall time: 977 ms


In [139]:
S[-1].mean()

np.float64(0.0)

In [140]:
%timeit S = mcs_simulation_cy((M, I))

961 ms ± 25.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
